src/main/java/org/wso2/carbon/esb/connector/jms/JMSPublisherPool.java

```diff
@@ -0,0 +1,143 @@
+     * @param deliveryMood             The message delivery mode for this message producer
+     * @param timeToLive               The message time to live in milliseconds
+     */
+    public JMSPublisherPool(String destination, String destinationType, String connectionFactoryName, int maxPoolSize,
+                            String javaNamingProviderUrl, String javaNamingFactoryInitial, String username,
+                            String password, String priority, String deliveryMood, String timeToLive) {
+        this.destination = destination;
+        this.destinationType = destinationType;
+        this.connectionFactoryName = connectionFactoryName;
+        this.maxSize = maxPoolSize;
+        this.javaNamingProviderUrl = javaNamingProviderUrl;
+        this.javaNamingFactoryInitial = javaNamingFactoryInitial;
+        this.publisherQueue = new ConcurrentLinkedQueue<>();
+        this.username = username;
+        this.password = password;
+        this.persistent = deliveryMood;
+        this.priority = priority;
+        this.timeToLive = timeToLive;
+    }
+
+    /**
+     * Will get the single publisher from the pool.
+     *
+     * @return The publisher to publish the message
+     * @throws JMSException    The JMXException
+     * @throws NamingException The NamingException
+     */
+    public JMSPublisher getPublisher() throws JMSException, NamingException {
+        JMSPublisher publisher = publisherQueue.poll();
+        if (publisher != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Returning an existing free publisher");
+            }
+            return publisher;
+        } else {
+            publisher = new JMSPublisher(destination, connectionFactoryName, destinationType,
+                    javaNamingProviderUrl, javaNamingFactoryInitial, username, password, priority, persistent, timeToLive);
+            if (log.isDebugEnabled()) {
+                log.debug("Created and returning a new publisher for destination");
+            }
+            return publisher;
+        }
+    }
+
+    /**
+     * Will release the publisher after the message published.
+     *
+     * @param publisher The publisher to be expired
+     */
+    public void releasePublisher(JMSPublisher publisher) {
+        if (publisherQueue.size() < maxSize) {
+            publisherQueue.add(publisher);
+            if (log.isDebugEnabled()) {
+                log.debug("Added publisher back to free pool.");
+            }
+        } else {
+            publisherQueue.poll().close();
+            publisherQueue.add(publisher);
+            if (log.isDebugEnabled()) {
+                log.debug("Destroying publisher because we have reached maximum size of publisher pool.");
+            }
+        }
+    }
+
+    /**
+     * Will clear all publishers from publisherPool.
+     *
+     * @throws JMSException The JMXException
+     */
+    public void close() throws JMSException {
```

Comment:
We are using this method from finalize to close all the connection when destroying the connector.